kubetest2/kubetest2-gce/deployer/up.go

```diff
@@ -79,6 +91,8 @@ func (d *deployer) verifyFlags() error {
 		return err
 	}
 
+	d.kubectl = filepath.Join(d.RepoRoot, "cluster", "kubectl.sh")
```

Comment:
perhaps we should move all the common logic, like buildEnv and deployer initialization behind a sync.Once initialization so that we don't need to do this in every step.

https://golang.org/pkg/sync/#Once

Reply:
Is it worth the code complexity? Implicitly, things like buildEnv and verifyFlags should have no effect if run more than once vs just once (I suppose this is idempotency).

Reply:
Yes, but there's no reason to run them multiple times.
`sync.Once` can be used pretty easily like so:
https://play.golang.org/p/1J9ghm_eI9t

shouldn't really add that much complexity.

Reply:
I guess I'm still not sure what is gained by using `sync.Once`, to me it seems to just add lines of code and complexity for a reader.

I don't see a potential situation where guaranteeing "run at most once" for any of the code is necessary. I doubt this code will ever be run concurrently in Go and therefore need to deal with races. The only situation (that I can see) in which we would gain some "safety" is the [completely pathological](https://unix.stackexchange.com/questions/38205/change-environment-of-a-running-process) case of a user editing the process's environment variables during its run.

If there's an argument to be made for performance, I don't think it matters because runtime is overwhelmingly dominated by kube-up.sh and company.

Reply:
Well, `buildEnv()` might be simple and innocuous but stuff like `verifyFlags()` and executing `kubectl.sh` in each step seems pretty unnecessary to me when we know it's explicitly a one-time thing. 
I guess it comes down to whether or not you think `d.initialize.Do(func() { err = d.init() })` is adding significant readability complexity as compared to `err = d.init()` .  I tend towards no, but I'll leave it to you.

@rancerbeta thoughts?

/lgtm
/hold

Reply:
It doesn't add too much more complexity. I think it would be a good idea to capture all your initialization at the front if possible. There's no saying what a test author may do in between each and removing that possibility is a good thing. Users are not always being logical.

Reply:
Here is a proposal for using `sync.Once`: https://github.com/michaelmdresser/test-infra/pull/1

I opted to remove the build-specific flag verification. I could bring it back and use another `sync.Once` for it, I'm definitely interested in discussing this choice.

I didn't put `buildEnv` behind a `sync.Once` because (please, correct me if I'm wrong) the only behavior that would make sense if it was behind a `sync.Once` is to build the environment and then store it as a slice in the deployer struct. However, because it is a slice, every time a script-caller needs to modify it, that caller has to copy the slice rather than just modify it. This seems like even more of a "relying on the future programmer" than what it is now.

I suppose `buildEnv` could also modify the os environment, but I think that is even less desirable because it is more opaque.

Reply:
on the sync.once thing, buildenv and verifyflags look like functions that should occur once per kubetest2 CLI invocation.  it might be sufficient to check the existence of a verifiedenv/verifiedflags variable without needing to do sync.once

afaik each of the stages is never run in parallel with each other. wdyt amit?

Reply:
After talking with Amit, I think I will implement this behavior alongside saving intermediate state of the deployer struct in the artifacts directory. I think this makes the most sense because that will introduce a need for a generic initialization step that should only be performed once.